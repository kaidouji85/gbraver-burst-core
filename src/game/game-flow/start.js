// @flow

import type {GameStateX} from '../../state/game-state';
import {GameFlowWithHistory} from './game-flow';

/**
 * ゲームフローを開始する
 *
 * @template X ゲームステートの効果
 * @param state ゲームステート
 * @return ゲームフロー
 */
export function start<X>(state: GameStateX<X>): GameFlowWithHistory<X> {
  return new GameFlowWithHistory([(state: any)], state);
}