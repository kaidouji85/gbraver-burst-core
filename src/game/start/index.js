// @flow
import type {GameState} from '../game-state';
import {createArmdozerGameState} from '../../armdozer/armdozer-game-state';
import {getFirstTurnPlayer} from './first-turn-payer';
import type {Player} from "../../player/player";
import {PhaseNameList} from "../phase-name";

/**
 * ゲームの初期状態を生成する
 *
 * @param player1 プレイヤー1
 * @param player2 プレイヤー2
 * @return ゲーム初期状態
 */
export function start(player1: Player, player2: Player): GameState[] {
  const players = [player1, player2]
    .map(v => ({
      playerId: v.playerId,
      armDozer: createArmdozerGameState(v.armDozer)
    }));
  const firstState: GameState = {
    players,
    phase: PhaseNameList.COMMAND_PHASE,
    activePlayerId: getFirstTurnPlayer(players[0], players[1])
  };
  return [firstState];
}