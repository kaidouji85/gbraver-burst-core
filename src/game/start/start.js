// @flow

import type {GameState} from "../../state/game-state";
import {startGame} from "../../effect/start-game";
import {gameStartInputCommand} from "../../effect/input-command";
import {upcastGameState as up} from "../../state/game-state";
import type {Player} from "../../player/player";
import {chain, start as startFlow} from '../state-updator/state-updator';

/**
 * ゲームの初期状態を生成する
 *
 * @param players ゲーム参加プレイヤー
 * @return ゲーム初期状態
 */
export function start(players: [Player, Player]): GameState[] {
  return startFlow(startGame(players))
    .to(chain(v => gameStartInputCommand(up(v))))
    .stateHistory;
}