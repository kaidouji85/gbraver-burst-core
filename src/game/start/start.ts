import { gameStartInputCommand } from "../../effect/input-command";
import { startGame } from "../../effect/start-game";
import type { Player } from "../../player/player";
import type { GameState } from "../../state/game-state";
import { startGameStateFlow } from "../game-state-flow";

/**
 * ゲームの初期状態を生成する
 *
 * @param players ゲーム参加プレイヤー
 * @return ゲーム初期状態
 */
export function start(players: [Player, Player]): GameState[] {
  return startGameStateFlow([startGame(players)])
    .add((state) => [gameStartInputCommand(state)])
    .toGameStateHistory();
}
