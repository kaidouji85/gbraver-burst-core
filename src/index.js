// @flow
export {ArmDozerIdList, ArmdozerAppearances, ArmDozers} from './master/armdozers';
export {GbraverBurstCore} from './game/index';

export type {Player, PlayerX, PlayerId} from './player/player';
export type {ArmDozerId, ArmdozerAppearance, Armdozer, ArmdozerX} from './player/armdozer';
export type {Burst, RecoverBattery, BuffPower, LightningBarrier, ContinuousAttack} from './player/burst';

export type {GameState, GameStateX} from './game/state/game-state';
export type {PlayerState, PlayerStateX} from './game/state/player-state';
export type {ArmdozerState, ArmdozerStateX} from './game/state/armdozer-state';
export type {ArmdozerEffect, EmptyArmdozerEffect, CorrectPower, TryReflect, ContinuousActivePlayer} from './game/state/armdozer-effect';

export type {Command, PlayerCommand, PlayerCommandX} from './command/command';
export type {BatteryCommand} from './command/battery';
export type {BurstCommand} from './command/burst';
export type {EmptyCommand} from './command/empty-command';

export type {Effect} from './effect/index';
export type {BatteryDeclaration} from './effect/battery-declaration/battery-declaration';
export type {Battle, BattleX} from './effect/battle/battle';
export type {BattleResult} from './effect/battle/result/battle-result';
export type {CriticalHit} from './effect/battle/result/critical-hit';
export type {Feint} from './effect/battle/result/feint';
export type {Guard} from './effect/battle/result/guard';
export type {Miss} from './effect/battle/result/miss';
export type {NormalHit} from './effect/battle/result/normal-hit';
export type {BurstEffect} from './effect/burst/burst-effect';
export type {GameEnd, GameEndX, GameEndResult} from './effect/game-end/game-end';
export type {GameOver, EvenMatch} from './game/end-judging/game-end-judging';
export type {InputCommand, Selectable, NoChoice} from './effect/input-command/input-command';
export type {Reflect, ReflectDamageEffect} from './effect/reflect/reflect';
export type {StartGame} from './effect/start-game/start-game';
export type {TurnChange} from './effect/turn-change/turn-change';
export type {UpdateRemainingTurn, EndArmdozerEffect} from './effect/update-remaning-turn/update-remaining-turn';
export type {RightItself} from './effect/right-itself/right-itself';

// TODO 以下は廃止予定
export {start} from './game/start/start';
export {progress} from './game/progress/index';
