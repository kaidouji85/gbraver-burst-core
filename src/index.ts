export { BatteryCommand, BatteryCommandSchema } from "./command/battery";
export {
  BurstCommand,
  BurstCommandSchema,
} from "./command/burst";
export {
  Command,
  CommandSchema,
  isCommandEqual,
} from "./command/command";
export {
  EmptyCommand,
  EmptyCommandSchema,
} from "./command/empty-command";
export {
  PilotSkillCommand,
  PilotSkillCommandSchema,
} from "./command/pilot-skill";
export { QuickCommand } from "./command/quick-command";
export {
  correctedBattery,
  totalBatteryCorrection,
} from "./effect/battery-correction";
export {
  BatteryDeclaration,
  BatteryDeclarationSchema,
} from "./effect/battery-declaration/battery-declaration";
export {
  Battle,
  BattleSchema,
  BattleX,
} from "./effect/battle/battle";
export { updateDefender } from "./effect/battle/players/update-defender";
export {
  BattleResult,
  battleResult,
  BattleResultSchema,
} from "./effect/battle/result/battle-result";
export {
  CriticalHit,
  CriticalHitSchema,
} from "./effect/battle/result/critical-hit";
export { Feint, FeintSchema } from "./effect/battle/result/feint";
export { Guard, GuardSchema } from "./effect/battle/result/guard";
export { Miss, MissSchema } from "./effect/battle/result/miss";
export {
  NormalHit,
  NormalHitSchema,
} from "./effect/battle/result/normal-hit";
export {
  BurstEffect,
  BurstEffectSchema,
} from "./effect/burst/burst-effect";
export { hasContinuousActive } from "./effect/continuous-active/has-continuous-active";
export {
  correctPower,
  hasHalveCorrectPower,
  totalCorrectPower,
} from "./effect/correct-power";
export {
  GameEnd,
  GameEndResult,
  GameEndResultSchema,
  GameEndX,
} from "./effect/game-end/game-end";
export { Effect } from "./effect/index";
export {
  InputCommand,
  NoChoice,
  Selectable,
} from "./effect/input-command/input-command";
export {
  PilotSkillEffect,
  PilotSkillEffectX,
} from "./effect/pilot-skill/pilot-skill-effect";
export { Reflect } from "./effect/reflect/reflect";
export { RightItself } from "./effect/right-itself/right-itself";
export { StartGame } from "./effect/start-game/start-game";
export { TurnChange } from "./effect/turn-change/turn-change";
export {
  EndArmdozerEffect,
  UpdateRemainingTurn,
} from "./effect/update-remaning-turn/update-remaining-turn";
export {
  EMPTY_ARMDOZER_EFFECT,
  EMPTY_CORRECT_POWER,
} from "./empty/amrdozer-effect";
export {
  EMPTY_ARMDOZER,
  EMPTY_ARMDOZER_STATE,
  EMPTY_BURST,
} from "./empty/armdozer";
export { EMPTY_BATTLE } from "./empty/battle";
export { EMPTY_GAME_STATE } from "./empty/game-state";
export { EMPTY_PILOT, EMPTY_PILOT_STATE } from "./empty/pilot";
export { EMPTY_PLAYER, EMPTY_PLAYER_STATE } from "./empty/player";
export { PlayerCommand, PlayerCommandX } from "./game/command/player-command";
export {
  EvenMatch,
  EvenMatchSchema,
  GameContinue,
  GameContinueSchema,
  GameEndJudging,
  GameEndJudgingSchema,
  GameOver,
  GameOverSchema,
} from "./game/end-judging/game-end-judging";
export {
  GbraverBurstCore,
  RestoreGbraverBurst,
} from "./game/gbraver-burst-core";
export {
  restoreGbraverBurst,
  startGbraverBurst,
} from "./game/gbraver-burst-core";
export { ArmDozerIds, ArmDozers } from "./master/armdozers";
export { PilotIds, Pilots } from "./master/pilots";
export {
  Armdozer,
  ArmDozerId,
  ArmDozerIdSchema,
  ArmdozerSchema,
  ArmdozerX,
} from "./player/armdozer";
export { Burst, BurstSchema } from "./player/burst";
export { BatteryLimitBreak, BatteryLimitBreakSchema } from "./player/burst/battery-limit-break";
export { BuffPower, BuffPowerSchema } from "./player/burst/buff-power";
export { ContinuousAttack, ContinuousAttackSchema } from "./player/burst/continuous-attack";
export { LightningBarrier, LightningBarrierSchema } from "./player/burst/lightning-barrier";
export { RecoverBattery, RecoverBatterySchema } from "./player/burst/recover-battery";
export {
  Pilot,
  PilotId,
  PilotIdSchema,
  PilotSchema,
  PilotX,
} from "./player/pilot";
export { BatteryBoostSkill, BatteryBoostSkillSchema } from "./player/pilot/battery-boost-skill";
export { BatteryEnchantmentSkill, BatteryEnchantmentSkillSchema } from "./player/pilot/battery-enchantment-skill";
export { BuffPowerSkill, BuffPowerSkillSchema } from "./player/pilot/buff-power-skill";
export { DamageHalvedSkill, DamageHalvedSkillSchema } from "./player/pilot/damage-halved-skill";
export { PilotSkill, PilotSkillSchema } from "./player/pilot/pilot-skill";
export { RecoverBatterySkill, RecoverBatterySkillSchema } from "./player/pilot/recover-battery-skill";
export {
  Player,
  PlayerId,
  PlayerIdSchema,
  PlayerSchema,
  PlayerX,
} from "./player/player";
export {
  ArmdozerEffect,
  ArmdozerEffectSchema,
} from "./state/armdozer-effect";
export { BatteryCorrection, BatteryCorrectionSchema } from "./state/armdozer-effect/battery-correction";
export { BatteryRecoverSkip, BatteryRecoverSkipSchema } from "./state/armdozer-effect/battery-recover-skip";
export { ContinuousActivePlayer, ContinuousActivePlayerSchema } from "./state/armdozer-effect/continuous-active-player";
export { CorrectPower, CorrectPowerSchema } from "./state/armdozer-effect/correct-power";
export { DamageHalved, DamageHalvedSchema } from "./state/armdozer-effect/damage-halved";
export { EffectPeriod, EffectPeriodSchema } from "./state/armdozer-effect/effect-period";
export { EmptyArmdozerEffect, EmptyArmdozerEffectSchema } from "./state/armdozer-effect/empty-armdozer-effect";
export { HalveCorrectPower, HalveCorrectPowerSchema } from "./state/armdozer-effect/halve-correct-power";
export { ReflectDamageEffect, ReflectDamageEffectSchema } from "./state/armdozer-effect/reflect-damage-effect";
export { parseSpecialPeriodEffect } from "./state/armdozer-effect/special-period-effect";
export { SpecialPeriodEffectSchema } from "./state/armdozer-effect/special-period-effect";
export { parseTryReflect } from "./state/armdozer-effect/try-reflect";
export { TryReflectSchema } from "./state/armdozer-effect/try-reflect";
export { TryReflect } from "./state/armdozer-effect/try-reflect";
export { parseTurnLimitEffect } from "./state/armdozer-effect/turn-limit-effect";
export { TurnLimitEffectSchema } from "./state/armdozer-effect/turn-limit-effect";
export {
  ArmdozerState,
  ArmdozerStateSchema,
  ArmdozerStateX,
  parseArmDozerState,
} from "./state/armdozer-state";
export { GameState, GameStateX } from "./state/game-state";
export {
  parsePilotState,
  PilotState,
  PilotStateSchema,
  PilotStateX,
} from "./state/pilot-state";
export {
  parsePlayerState,
  PlayerState,
  PlayerStateSchema,
  PlayerStateX,
} from "./state/player-state";
export { isPlayerDeath } from "./state/player-state/is-player-death";
