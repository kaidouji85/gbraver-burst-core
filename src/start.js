// @flow
import type {Player} from "./player/player";
import type {GameState} from "./game-state/game-state";
import {startGame} from "./effect/start-game/index";
import {inputCommand} from "./effect/input-command/index";

/**
 * ゲームの初期状態を生成する
 *
 * @param player1 プレイヤー1
 * @param player2 プレイヤー2
 * @return ゲーム初期状態
 */
export function start(player1: Player, player2: Player): GameState[] {
  const initialState = startGame(player1, player2);
  return [
    inputCommand(initialState)
  ];
}