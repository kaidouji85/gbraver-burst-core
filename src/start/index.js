// @flow
import type {GameState} from '../game-state/index';
import {getFirstTurnPlayer} from './first-turn-payer';
import type {Player} from "../player/player";
import {PhaseNameList} from "../phase/phase-name";
import {createOpenPlayerState} from "../game-state/open-player-state";

/**
 * ゲームの初期状態を生成する
 *
 * @param player1 プレイヤー1
 * @param player2 プレイヤー2
 * @return ゲーム初期状態
 */
export function start(player1: Player, player2: Player): GameState {
  const openPlayerStateList = [player1, player2].map(v => createOpenPlayerState(v));
  const secretPlayerStateList = [player1, player2]
    .map(v => ({
      playerId: v.playerId,
      lastCommand: {type: 'EMPTY_COMMAND'}
    }));

  return {
    steps: [{
      openState: {
        players: openPlayerStateList,
        phase: PhaseNameList.COMMAND_PHASE,
        activePlayerId: getFirstTurnPlayer(openPlayerStateList[0], openPlayerStateList[1])
      },
      secretState: {
        players: secretPlayerStateList
      }
    }]
  };
}