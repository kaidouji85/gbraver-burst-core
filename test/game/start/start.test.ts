import type { Player } from "../../../src";
import { EMPTY_ARMDOZER, EMPTY_PILOT } from "../../../src";
import { start } from "../../../src/game/start/start";

const player1: Player = {
  playerId: "player1",
  pilot: EMPTY_PILOT,
  armdozer: { ...EMPTY_ARMDOZER, speed: 1600 },
};
const player2: Player = {
  playerId: "player2",
  pilot: EMPTY_PILOT,
  armdozer: { ...EMPTY_ARMDOZER, speed: 2000 },
};

test("ゲームを正しく開始できる", () => {
  const result = start([player1, player2]);
  expect(result).toMatchSnapshot();
});
