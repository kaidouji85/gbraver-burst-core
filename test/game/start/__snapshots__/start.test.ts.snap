// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ゲームを正しく開始できる 1`] = `
[
  {
    "activePlayerId": "player2",
    "effect": {
      "name": "StartGame",
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
  {
    "activePlayerId": "player2",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player1",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "player2",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 1600,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player1",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "player2",
      },
    ],
  },
]
`;
